src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -8,46 +8,48 @@
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Serialization;
 using Microsoft.CodeAnalysis.SolutionCrawler;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     /// <summary>
-    /// primary workspace for remote host. no one except solution service can update this workspace
+    /// Workspace created by the remote host that mirrors the corresponding client workspace.
     /// </summary>
-    internal class RemoteWorkspace : Workspace
+    internal sealed class RemoteWorkspace : Workspace
     {
         private readonly ISolutionCrawlerRegistrationService? _registrationService;
 
         // guard to make sure host API doesn't run concurrently
         private readonly object _gate = new object();
 
+        private readonly SemaphoreSlim _solutionSynchronizationGate = new SemaphoreSlim(initialCount: 1);
+
+        // the last solution for the the primary branch fetched from the client
+        private volatile Tuple<Checksum, Solution>? _primaryBranchSolutionWithChecksum;
+
+        // the last solution requested by a service:
```

Comment:
use `///` comments.  Also please doc the _gate and semaphore to say what they're protecting given that we have two of them.